File: tests/fixtures/inputs/quotations/oqo/tools/validate_quote.py
#!/usr/bin/env python3
"""
validate_quote.py – Validate a quote JSON file against the OQO schema.

Usage:
    python validate_quote.py <quote.json> <schema.json>

Requires the `jsonschema` package (install via `pip install jsonschema`).
"""
import json
import sys
from pathlib import Path

try:
    from jsonschema import Draft202012Validator
except ImportError:
    print("Error: The 'jsonschema' package is required. Install it via 'pip install jsonschema'.")
    sys.exit(1)


def load_json(path: Path):
    with path.open('r', encoding='utf-8') as f:
        return json.load(f)


def main():
    if len(sys.argv) != 3:
        print(__doc__)
        sys.exit(1)
    quote_path = Path(sys.argv[1])
    schema_path = Path(sys.argv[2])
    if not quote_path.exists() or not schema_path.exists():
        print(f"Error: File(s) not found: {quote_path}, {schema_path}")
        sys.exit(1)
    quote_data = load_json(quote_path)
    schema_data = load_json(schema_path)
    validator = Draft202012Validator(schema_data)
    errors = sorted(validator.iter_errors(quote_data), key=lambda e: e.path)
    if not errors:
        print("PASS: Quote is valid according to the schema.")
    else:
        print(f"FAIL: Quote is invalid. Found {len(errors)} error(s):")
        for idx, error in enumerate(errors, 1):
            path = '.'.join(str(p) for p in error.path) or 'root'
            print(f" {idx}. [{path}] {error.message}")
        sys.exit(1)


if __name__ == '__main__':
    main()

File: tests/fixtures/inputs/quotations/oqo/tools/validate_quote.js
/*
 * validate_quote.js – Client‑side validator for OQO JSON objects using Ajv.
 *
 * This script loads the Ajv 8 validator from a CDN and validates a given JSON
 * object against the OQO schema.  It can be used in a browser or bundled into
 * your own application.  See the demo page for an example of how to use it.
 */

// Load Ajv from a CDN (if not already loaded).  This returns a Promise that
// resolves with an Ajv instance.
function loadAjv() {
  if (window.Ajv) {
    return Promise.resolve(window.Ajv);
  }
  return new Promise((resolve, reject) => {
    const script = document.createElement('script');
    script.src = 'https://cdn.jsdelivr.net/npm/ajv@8/dist/ajv.min.js';
    script.onload = () => resolve(window.Ajv);
    script.onerror = () => reject(new Error('Failed to load Ajv')); 
    document.head.appendChild(script);
  });
}

/**
 * Validate a quote JSON object against the provided schema.
 *
 * @param {Object} quoteData The quote to validate
 * @param {Object} schemaData The JSON Schema definition
 * @returns {Promise<{valid: boolean, errors: Array}>}
 */
async function validateQuote(quoteData, schemaData) {
  const Ajv = await loadAjv();
  const ajv = new Ajv({ strict: false, allErrors: true });
  const validate = ajv.compile(schemaData);
  const valid = validate(quoteData);
  return { valid, errors: validate.errors || [] };
}

// Example usage (uncomment to test in browser console):
// fetch('open-schema/oqo.json').then(r => r.json()).then(async schema => {
//   const quote = await fetch('sample_quote.json').then(r => r.json());
//   const { valid, errors } = await validateQuote(quote, schema);
//   if (valid) {
//     console.log('Quote is valid');
//   } else {
//     console.error('Quote is invalid:', errors);
//   }
// });
